src/runtime_src/xdp/profile/plugin/noc/noc_plugin.cpp

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright (C) 2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#define XDP_SOURCE
+
+#include "xdp/profile/plugin/noc/noc_plugin.h"
+#include "xdp/profile/writer/noc/noc_writer.h"
+
+#include "core/common/system.h"
+#include "core/common/time.h"
+#include "core/common/config_reader.h"
+#include "core/include/experimental/xrt-next.h"
+
+#include <boost/algorithm/string.hpp>
+
+namespace xdp {
+
+  NOCProfilingPlugin::NOCProfilingPlugin() 
+      : XDPPlugin(), mKeepPolling(true)
+  {
+    db->registerPlugin(this);
+   
+    // Just like HAL and power profiling, go through devices 
+    //  that exist and open a file for each
+    uint64_t index = 0;
+    void* handle = xclOpen(index, "/dev/null", XCL_INFO);
+    while (handle != nullptr) {
+      // Determine the name of the device
+      struct xclDeviceInfo2 info;
+      xclGetDeviceInfo2(handle, &info);
+      std::string deviceName = std::string(info.mName);
+      mDevices.push_back(deviceName);
+
+      std::string outputFile = "noc_profile_" + deviceName + ".csv"; 
+      writers.push_back(new NOCProfilingWriter(outputFile.c_str(),
+			    deviceName.c_str(), index));
+      (db->getStaticInfo()).addOpenedFile(outputFile.c_str(), "NOC_PROFILE") ;
+
+      // Move on to next device
+      xclClose(handle);
+      ++index;
+      handle = xclOpen(index, "/dev/null", XCL_INFO);
+    }
+
+    // Get polling interval (in msec)
+    mPollingInterval = xrt_core::config::get_noc_profile_interval_ms();
+
+    // Start the NOC profiling thread
+    mPollingThread = std::thread(&NOCProfilingPlugin::pollNOCCounters, this);
```

Comment:
Do not use obfuscated API.
Furthermore, you should probably use `std::async` instead.
```suggestion
    mPollingThread = std::thread([&] { pollNOCCounters(); });
```

Reply:
not addresses in this PR